### PR TITLE
TOMEE-4351 - Jakarta Security 3.0 - OpenIdAuthenticationMechanismDefinition implementation

### PR DESCRIPTION
**Pull request by jungm**
Opening as a draft PR for now as it's still WIP

### Doesn't work/WIP:
- Token expiration checks (`@AutoApplySession` is used right now on the auth mechanism, I've mainly done this for testing purposes and will likely drop it)
- redirectToOriginalResource=true, isn't implemented yet
- Large parts of the code aren't covered by unit/integration tests

### What does work:
- Authentication against Keycloak (Both in a real app and in `examples/openid-security` tests) and https://github.com/mitreid-connect/OpenID-Connect-Java-Spring-Server (used in security TCK)
- Passes TCK app-openid and app-openid2, app-openid3 tests redirectToOriginalResource which is not implemented yet:
  - ```
    [INFO] Reactor Summary for Jakarta Security TCK - main 3.0.1:
    [INFO] 
    [INFO] Jakarta Security TCK - main ........................ SUCCESS [  2.672 s]
    [INFO] common ............................................. SUCCESS [  0.699 s]
    [INFO] app-openid ......................................... SUCCESS [  6.441 s]
    [INFO] app-openid2 ........................................ SUCCESS [ 16.031 s]
    [INFO] app-openid3 ........................................ FAILURE [  6.057 s]
    [INFO] ------------------------------------------------------------------------
    [INFO] BUILD FAILURE
    [INFO] ------------------------------------------------------------------------
    ```

I've been working on this on and off for a couple of weeks now, with this PR getting quite big by now. So I'd highly appreciate if someone can take a look and give some feedback 🙂

### CODE REVIEW

**Comment by jungm** [comment]
redirectToOriginalResource=true works now and all openid modules of the TCK are passing:
```
[INFO] Reactor Summary for Jakarta Security TCK - main 3.0.1:
[INFO] 
[INFO] Jakarta Security TCK - main ........................ SUCCESS [  2.725 s]
[INFO] common ............................................. SUCCESS [  0.748 s]
[INFO] app-openid ......................................... SUCCESS [  6.556 s]
[INFO] app-openid2 ........................................ SUCCESS [ 15.742 s]
[INFO] app-openid3 ........................................ SUCCESS [ 14.100 s]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  39.986 s
[INFO] Finished at: 2024-06-02T17:24:15+02:00
[INFO] ------------------------------------------------------------------------
```

**Comment by jungm** [comment]
Thanks @rzo1 for triggering a full build: https://ci-builds.apache.org/job/Tomee/view/tomee-10.x/job/pull-request-manual/106/

All tests are passing :slightly_smiling_face:

**Comment by tandraschko** [comment]
looks really good from a short review and tests are running; +1

great work!

**Review by rzo1 — COMMENTED** [review]
Looks really good from a short review.

File: tomee/tomee-security/pom.xml [hunk 1]
```diff
@@ -94,13 +94,25 @@
       <artifactId>bval-jsr</artifactId>
     </dependency>
 
+    <dependency>
+      <groupId>org.bitbucket.b_c</groupId>
+      <artifactId>jose4j</artifactId>
+      <version>0.9.6</version>
```
[the remark this answers is not in the document]
Comment (jungm) [reply tomee/tomee-security/pom.xml:100]: yeah absolutely makes sense

File: examples/security-openid/pom.xml [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+    <packaging>war</packaging>
+
+    <name>TomEE :: Examples :: Jakarta Security OpenID Connect</name>
+
+    <properties>
+        <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
+        <version.tomee>10.0.0-M2-SNAPSHOT</version.tomee>
+        <version.microprofile.config>3.0.2</version.microprofile.config>
+        <junit.jupiter.version>5.8.2</junit.jupiter.version>
+        <arquillian.version>1.7.0.Final</arquillian.version>
```
[the remark this answers is not in the document]
Comment (jungm) [reply examples/security-openid/pom.xml:33]: 1.7.1.Final apparently https://github.com/apache/tomee/commit/92c924671d0cb1b1029e6b046a37b101efbc1e8a, my branch just didn't contain this commit yet that's why I used 1.7.0.Final
I'll rebase/merge and update it

[the remark this answers is not in the document]
Comment (rzo1) [reply examples/security-openid/pom.xml:33]: https://github.com/apache/tomee/commit/1e1df457855c37420e231a1a825052941dc2af1d

File: examples/security-openid/pom.xml [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+    <version>10.0.0-M2-SNAPSHOT</version>
+    <packaging>war</packaging>
+
+    <name>TomEE :: Examples :: Jakarta Security OpenID Connect</name>
+
+    <properties>
+        <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
+        <version.tomee>10.0.0-M2-SNAPSHOT</version.tomee>
+        <version.microprofile.config>3.0.2</version.microprofile.config>
+        <junit.jupiter.version>5.8.2</junit.jupiter.version>
```
[the remark this answers is not in the document]
Comment (jungm) [reply examples/security-openid/pom.xml:32]: No reason as to why i chose this other than all the other examples being on 5.8.2 as well (even on main)

[the remark this answers is not in the document]
Comment (rzo1) [reply examples/security-openid/pom.xml:32]: Yep. Let's leave it as is and update everything later ;-)

**Comment by jungm** [comment]
what really needs a closer look and maybe needs to be discussed:

- JWT Validation is using jose.4.j, this introduces a new dependency in all tomee flavours (wasn't in webprofile before). Maybe it needs to be added in some notice file?
- Spec mentions a special variable that can be used in the annotation: ${baseURL}, I implemented this with producing an @Named String
- I built a basic delegate in OpenIdAuthenticationMechanismDefinitionDelegate that automatically resolves the configuration from the openid provider
- SavedRequest (originally from @LoginToContinue) has been rewritten so I can serialize it for use in cookies
- Spec is ambiguous on how to handle subjectTypeSupported, idTokenSigningAlgorithmsSupported and responseTypeSupported (See CompositeOpenIdProviderMetadata). A user can override these, but it's not obvious if that has been done or not. I handled these the same way soteria does, but it's probably worth a spec issue in the future?
- Requests to openid provider are done using JAX-RS Client, maybe we want to use something else in TomEE? Really the only reason I chose this was because it's convenient 

(See https://lists.apache.org/thread/sghf41f1z75gpnhpf236o1lrj1sl4vr8 for whole thread on mailing list)

**Review by rzo1 — APPROVED** [review]

**Comment by rzo1** [comment]
Thanks @jungm

**Pull request merged**